csharp/src/Ice/TraceUtil.cs

```diff
@@ -2,315 +2,190 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 //
 
-using ZeroC.Ice;
+using System;
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.Globalization;
+using System.IO;
 
-namespace IceInternal
+namespace ZeroC.Ice
 {
     internal static class TraceUtil
     {
-        internal static void TraceSend(Communicator communicator, IList<System.ArraySegment<byte>> buffer)
+        internal static void TraceFrame(Communicator communicator, Span<byte> header, OutgoingRequestFrame frame) =>
+            TraceRequest(
+                "sending request",
+                communicator,
+                header,
+                frame.Identity,
+                frame.Facet,
+                frame.Operation,
+                frame.IsIdempotent,
+                frame.Context,
+                frame.Encoding);
+
+        internal static void TraceFrame(Communicator communicator, Span<byte> header, OutgoingResponseFrame frame) =>
+            TraceResponse(
+                "sending response",
+                communicator,
+                header,
+                frame.ReplyStatus,
+                frame.Encoding);
+
+        internal static void TraceFrame(Communicator communicator, Span<byte> header, IncomingRequestFrame frame) =>
+            TraceRequest(
+                "received request",
+                communicator,
+                header,
+                frame.Identity,
+                frame.Facet,
+                frame.Operation,
+                frame.IsIdempotent,
+                frame.Context,
+                frame.Encoding);
+
+        internal static void TraceFrame(Communicator communicator, Span<byte> header, IncomingResponseFrame frame) =>
+            TraceResponse(
+                "received response",
+                communicator,
+                header,
+                frame.ReplyStatus,
+                frame.Encoding);
+
+        private static void TraceRequest(
+            string heading,
```

Comment:
I finding `heading` confusing, especially when another parameter is named `header`.
What about: `traceMessagePrefix` ?

Reply:
fixed